src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -0,0 +1,174 @@
+                                train(trainingJob, listener);
+                            },
+                            exception -> {
+                                // Serialization failed. Let listener handle the exception, but free up resources.
+                                jobCount.decrementAndGet();
+                                semaphore.release();
+                                logger.error("Unable to initialize model serialization: " + exception.getMessage());
+                                listener.onFailure(exception);
+                            }
+                    ),
+                    false
+            );
+        } catch (IOException ioe) {
+            jobCount.decrementAndGet();
+            semaphore.release();
+            throw ioe;
+        }
+    }
+
+    private void train(TrainingJob trainingJob, ActionListener<IndexResponse> listener) {
+        // Attempt to submit job to training thread pool. On failure, release the resources and serialize the failure.
+        try {
+            threadPool.executor(TRAIN_THREAD_POOL).execute(() -> {
+                try {
+                    trainingJob.run();
+                    serializeModel(trainingJob, listener, true);
+                } catch (IOException e) {
+                    logger.error("Unable to serialize model \"" + trainingJob.getModelId() + "\": " + e.getMessage());
+                    listener.onFailure(e);
+                } catch (Exception e) {
+                    logger.error("Unable to complete training for \"" + trainingJob.getModelId() + "\": "
+                            + e.getMessage());
+                    listener.onFailure(e);
+                } finally {
+                    jobCount.decrementAndGet();
+                    semaphore.release();
+                }
+            });
+        } catch (RejectedExecutionException ree) {
+            logger.error("Unable to train model \"" + trainingJob.getModelId() + "\": " + ree.getMessage());
+
+            ModelMetadata modelMetadata = trainingJob.getModel().getModelMetadata();
+            modelMetadata.setState(ModelState.FAILED);
+            modelMetadata.setError(ree.getMessage());
+
+            try {
+                serializeModel(trainingJob, listener, true);
+            } catch (IOException ioe) {
+                logger.error("Unable to serialize the failure for model \"" + trainingJob.getModelId() + "\": " + ioe);
+                listener.onFailure(ioe);
+            } finally {
+                jobCount.decrementAndGet();
+                semaphore.release();
+            }
+        }
+    }
+
+    private void serializeModel(TrainingJob trainingJob, ActionListener<IndexResponse> listener, boolean update)
+            throws IOException {
+        if (update) {
+            modelDao.update(trainingJob.getModelId(), trainingJob.getModel(), listener);
+        } else {
+            modelDao.put(trainingJob.getModelId(), trainingJob.getModel(), listener);
```

Comment:
What is difference between put and update?

Reply:
Put will fail if the model id is not null and the model id exists.